docs/style.md

```diff
@@ -109,6 +109,44 @@ Note that for some *big* `Copy` types, notably `CryptoHash`, we sometimes use
 references for performance reasons. As a rule of thumb, `T` is considered *big* if
 `size_of::<T>() > 2 * size_of::<usize>()`.
 
+### Prefer for loops over `for_each` and `try_for_each` methods
```

Comment:
I would love for this section to be more general and `for` vs `for_each` be just an example that the rule applies to. The same rationale applies to e.g. `{Option, Result}::and_then` as well, for example.

I was going to suggest something like:

> # Prefer native language constructs over methods taking a closure

but this wording doesn’t really work in cases like `map_err` which are much nicer with a method, and is probably way too general anyway. I wonder if there is any way to word this such that it only applies to the situations where a method would be a negative value…

Reply:
I think there's also value in this being very specific: `.for_each` is the case where indeed this is very clear cut, generalizing might make this more obscure.

Reply:
Maybe we shouldn’t have this in style after all?  It is subjective after all; I wouldn’t advocate outright ban of `for_each`.  For example, I kinda prefer for_each version below (and I initially did not change it to a loop):

```
-        peers.iter().for_each(|p| update.delete::<schema::Peers>(p));
+        for p in peers {
+            update.delete::<schema::Peers>(p);
+        }
```

And regarding `and_then` (and also example in rust-analyzer has in its style guide), the below is much more readable than alternatives:

```rust
            Ok(accumulator)
                .and_then(|accumulator| safe_add_balance(accumulator, amount))
                .and_then(|accumulator| safe_add_balance(accumulator, locked))
                .map_err(|_| RuntimeError::UnexpectedIntegerOverflow)
```

Even though `Ok(accumulator)` is pointless in a sense, it improves readability.

Reply:
I do think it's very valuable to have some form of this advice in the style guide: there's bimodal distribution of how people write code, there are folks who do mostly imperative style, and folks who do mostly functional style. I think it's valuable if we have a general culture of "we are mostly imperative" or "we are mostly functional". 

There's specific problem that in OSS projects some contributors love submitting PR with highly high-order code, and its useful in this cases to point to the style guide and say "look, we appreciate the beauty of functional composition, but we explicitly prefer first-order code".

At the same time, this definitely should not be a "law" rule, more of the guideline, so some softening of the language in the styleguide would be good. As ra puts this, `This rule is more "soft" than others`. 

In this sense, I actually like being specific about `.for_each` -- it is the more clear-cut case, but still sets the overall "mostly imperative" culture. 

>For example, I kinda prefer for_each version below (and I initially did not change it to a loop):

For me, I would say that I also prefer the `.for_each` version in isolation, but, if I think about surrounding context (`for_each` is used rarely elsewhere) and evolution (the closure might get bigger, or it might get an `?`), this tips the scale in favor of `for` for me.

Reply:
OK, I dunno.  I’m just gonna merge this as is.  Someone smarter may expand the wording if they have an idea how.